### PR TITLE
Add Kokkos::create_mirror[_view] WithoutInitializing for Container types

### PR DESCRIPTION
**Pull request by masterleinad**
Related to #4794, in continuation of #4486.

`StaticCrsGraph` also provides `create_mirror[_view]` but with somewhat different semantics so I decided to omit it here.

### CODE REVIEW

**Comment by stanmoore1** [comment]
Thanks @masterleinad!

File: containers/src/Kokkos_OffsetView.hpp [hunk 1]
```diff
@@ -1883,160 +1883,145 @@ struct MirrorOffsetType {
...
+      Kokkos::create_mirror(arg_prop..., src.view()), src.begins());
+}
+
+template <class Space, class T, class... P, class... I>
+inline typename Kokkos::Impl::MirrorOffsetType<Space, T, P...>::view_type
+create_mirror(const Space&,
+              const Kokkos::Experimental::OffsetView<T, P...>& src,
+              const I&... arg_prop) {
+  return typename Kokkos::Impl::MirrorOffsetType<Space, T, P...>::view_type(
+      Kokkos::view_alloc(src.label(), arg_prop...), src.layout(),
```
Comment (PhilMiller) [inline containers/src/Kokkos_OffsetView.hpp:1901]: Is switching from constructing `array_layout` to passing `src.layout()` also expected to handle non-unit strides, where it didn't previously?

Comment (masterleinad) [reply containers/src/Kokkos_OffsetView.hpp:1901]: I would need to evaluate that.

File: containers/src/Kokkos_DynamicView.hpp [hunk 2]
```diff
@@ -540,34 +544,202 @@ class DynamicView : public Kokkos::ViewTraits<DataType, P...> {
...
+struct MirrorDynamicViewType {
+  // The incoming view_type
+  using src_view_type = typename Kokkos::Experimental::DynamicView<T, P...>;
+  // The memory space for the mirror view
+  using memory_space = typename Space::memory_space;
+  // Check whether it is the same memory space
+  enum {
+    is_same_memspace =
+        std::is_same<memory_space, typename src_view_type::memory_space>::value
+  };
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply containers/src/Kokkos_DynamicView.hpp:592]: I prefer just doing the same as in the respective classes for other View types and got through all the unnamed lambdas later.

File: containers/src/Kokkos_OffsetView.hpp (hunk 1 above)
Comment (masterleinad) [reply containers/src/Kokkos_OffsetView.hpp:1901]: The constructor called has the same implementation as the one for `Kokkos::View` so the answer should be"Yes, it allows non-unit strides".

**Review by PhilMiller — APPROVED** [review]

File: containers/src/Kokkos_DynRankView.hpp [hunk 3]
```diff
@@ -1952,74 +1951,113 @@ inline typename DynRankView<T, P...>::HostMirror create_mirror(
...
+    const Space&, const Kokkos::DynRankView<T, P...>& src) {
+  return typename Impl::MirrorDRVType<Space, T, P...>::view_type(
+      src.label(), Impl::reconstructLayout(src.layout(), src.rank()));
 }
 
-// Create a mirror in a new space (specialization for different space)
 template <class Space, class T, class... P>
 typename Impl::MirrorDRVType<Space, T, P...>::view_type create_mirror(
-    const Space&, const Kokkos::DynRankView<T, P...>& src,
+    Kokkos::Impl::WithoutInitializing_t wi, const Space&,
```
[the remark this answers is not in the document]
Comment (crtrott) [reply containers/src/Kokkos_DynRankView.hpp:1979]: nvm we did this in the main create_mirror too ...

File: containers/src/Kokkos_DynRankView.hpp [hunk 4]
```diff
@@ -1952,74 +1951,113 @@ inline typename DynRankView<T, P...>::HostMirror create_mirror(
...
+    const Kokkos::DynRankView<T, P...>& src,
     typename std::enable_if<std::is_same<
         typename ViewTraits<T, P...>::specialize, void>::value>::type* =
         nullptr) {
   return typename Impl::MirrorDRVType<Space, T, P...>::view_type(
-      src.label(), Impl::reconstructLayout(src.layout(), src.rank()));
+      Kokkos::view_alloc(std::string(src.label()).append("_mirror"), wi),
```
[the remark this answers is not in the document]
Comment (crtrott) [reply containers/src/Kokkos_DynRankView.hpp:1985]: nvm

File: containers/src/Kokkos_DynRankView.hpp [hunk 5]
```diff
@@ -1952,74 +1951,113 @@ inline typename DynRankView<T, P...>::HostMirror create_mirror(
...
 template <class Space, class T, class... P>
-typename Impl::MirrorDRViewType<Space, T, P...>::view_type create_mirror_view(
-    const Space&, const Kokkos::DynRankView<T, P...>& src,
-    typename std::enable_if<
-        !Impl::MirrorDRViewType<Space, T, P...>::is_same_memspace>::type* =
-        nullptr) {
-  return typename Impl::MirrorDRViewType<Space, T, P...>::view_type(
-      src.label(), Impl::reconstructLayout(src.layout(), src.rank()));
+inline auto create_mirror_view(Kokkos::Impl::WithoutInitializing_t wi,
+                               const Space& space,
```
[the remark this answers is not in the document]
Comment (crtrott) [reply containers/src/Kokkos_DynRankView.hpp:2058]: nvm

**Review by crtrott — APPROVED** [review]

File: containers/src/Kokkos_DynRankView.hpp [hunk 6]
```diff
@@ -1936,13 +1936,12 @@ struct MirrorDRVType {
 
 }  // namespace Impl
 
+// Create a mirror in host space
 template <class T, class... P>
 inline typename DynRankView<T, P...>::HostMirror create_mirror(
     const DynRankView<T, P...>& src,
-    typename std::enable_if<
-        std::is_same<typename ViewTraits<T, P...>::specialize, void>::value &&
-        !std::is_same<typename Kokkos::ViewTraits<T, P...>::array_layout,
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply containers/src/Kokkos_DynRankView.hpp:1944]: Yes.

File: containers/src/Kokkos_DynRankView.hpp [hunk 7]
```diff
@@ -1952,74 +1951,113 @@ inline typename DynRankView<T, P...>::HostMirror create_mirror(
...
+// Create a mirror in a new space
+template <
+    class Space, class T, class... P,
+    typename Enable = std::enable_if_t<
+        Kokkos::is_space<Space>::value &&
+        std::is_same<typename ViewTraits<T, P...>::specialize, void>::value>>
+typename Impl::MirrorDRVType<Space, T, P...>::view_type create_mirror(
+    const Space&, const Kokkos::DynRankView<T, P...>& src) {
+  return typename Impl::MirrorDRVType<Space, T, P...>::view_type(
+      src.label(), Impl::reconstructLayout(src.layout(), src.rank()));
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply containers/src/Kokkos_DynRankView.hpp:1974]: `Space` is (allowed to be) a memory space. We can discuss if we want to specialize for the case that an execution space is provided but I would prefer to do that uniformly with the other `create_mirror` overloads in a separate pull request

**Pull request merged**